cudamapper/src/matcher_gpu.cuh

```diff
@@ -108,13 +148,196 @@ void compute_anchor_starting_indices(
+/// \param compound_key_positions_in_reads_d the array to be filled with compund keys, the size of this array has to be equal to the last element of anchor_starting_indices
+/// \param n_anchors the size of the anchors array
+/// \param anchor_starting_indices_d the array of starting indices of the set of anchors for each unique representation of the query index (representations with no match in target will have the same starting index as the last matching representation)
+/// \param query_starting_index_of_each_representation_d the starting index of a representation in query_read_ids and query_positions_in_read
+/// \param found_target_indices_d the found matches in the array of unique target representation for each unique representation of query index
+/// \param target_starting_index_of_each_representation_d the starting index of a representation in target_read_ids and target_positions_in_read
+/// \param n_query_representations the size of the query_starting_index_of_each_representation_d and found_target_indices_d arrays, ie. the number of unique representations in the query index
+/// \param query_read_ids the array of read ids of the (read id, position)-pairs in query index
+/// \param query_positions_in_read the array of positions of the (read id, position)-pairs in query index
+/// \param target_read_ids the array of read ids of the (read id, position)-pairs in target index
+/// \param target_positions_in_read the array of positions of the (read id, position)-pairs in target index
+/// \param smallest_query_read_id smallest read_id in query index
+/// \param smallest_target_read_id smallest read_id in target index
+/// \param number_of_target_reads number of read_ids in taget index
+/// \param max_basepairs_in_target_reads number of basepairs in longest read in target index
+/// \tparam ReadsKeyT type of compound_key_read_ids_d, has to be integral
+/// \tparam PositionsKeyT type of compound_key_positions_in_reads_d, has to be integral
+template <typename ReadsKeyT, typename PositionsKeyT>
+__global__ void generate_anchors_kernel(
+    Anchor* const anchors_d,
+    ReadsKeyT* const compound_key_read_ids_d,
+    PositionsKeyT* const compound_key_positions_in_reads_d,
+    const int64_t n_anchors,
+    const int64_t* const anchor_starting_index_d,
+    const std::uint32_t* const query_starting_index_of_each_representation_d,
+    const std::int64_t* const found_target_indices_d,
+    const int32_t n_query_representations,
+    const std::uint32_t* const target_starting_index_of_each_representation_d,
+    const read_id_t* const query_read_ids,
+    const position_in_read_t* const query_positions_in_read,
+    const read_id_t* const target_read_ids,
+    const position_in_read_t* const target_positions_in_read,
+    const read_id_t smallest_query_read_id,
+    const read_id_t smallest_target_read_id,
+    const read_id_t number_of_target_reads,
```

Comment:
I'm not sure: should `number_of_target_reads` really be a `read_id_t`?

Reply:
Good point. `number_of_target_reads` is not a `read_id`, so in that sense it's misleading. On the other hand, size of `read_id_t` is always going to be right to hold `number_of_target_reads`, but I agree it's a bit hacky.
An alternative would be to make it `int32_t` as we "know" we are never going to have more reads than that (and then we do and it causes a whole bunch of problems).
I'd personally leave it as it is because otherwise it make the rest of the code more complicated.
I think we should find a more general solution for this type of problems, but that leads to whole "safe math in C++" discussion